File: streampark-common/src/main/scala/org/apache/streampark/common/util/CompletableFutureUtils.scala
/*
 * Licensed to the Apache Software Foundation (ASF) under one or more
 * contributor license agreements.  See the NOTICE file distributed with
 * this work for additional information regarding copyright ownership.
 * The ASF licenses this file to You under the Apache License, Version 2.0
 * (the "License"); you may not use this file except in compliance with
 * the License.  You may obtain a copy of the License at
 *
 *    http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 */

package org.apache.streampark.common.util

import java.util.concurrent.{Callable, CompletableFuture, ScheduledThreadPoolExecutor, ThreadFactory, TimeoutException, TimeUnit}
import java.util.function.{BiConsumer, Consumer, Function => JavaFunc}

object CompletableFutureUtils {

  private[this] class DaemonThreadFactory extends ThreadFactory {
    override def newThread(r: Runnable): Thread = {
      val t = new Thread(r)
      t.setDaemon(true)
      t.setName("CompletableFutureDelayScheduler")
      t
    }
  }

  private[this] lazy val completableDelayer: ScheduledThreadPoolExecutor = {
    val delayer = new ScheduledThreadPoolExecutor(1, new DaemonThreadFactory())
    delayer.setRemoveOnCancelPolicy(true)
    delayer
  }

  private[this] def setTimeout[T](timeout: Long, unit: TimeUnit): CompletableFuture[T] = {
    val result = new CompletableFuture[T]
    completableDelayer.schedule(
      new Callable[Boolean] {
        override def call(): Boolean =
          result.completeExceptionally(new TimeoutException)
      },
      timeout,
      unit)
    result
  }

  def supplyTimeout[T](
      future: CompletableFuture[T],
      timeout: Long,
      unit: TimeUnit,
      handle: JavaFunc[T, T],
      exceptionally: JavaFunc[Throwable, T]): CompletableFuture[T] = {
    future
      .applyToEither(setTimeout(timeout, unit), handle)
      .exceptionally(exceptionally)
      .whenComplete(new BiConsumer[T, Throwable]() {
        override def accept(t: T, u: Throwable): Unit = {
          if (!future.isDone) {
            future.cancel(true)
          }
        }
      })
  }

  /**
   * Callback handler when the future is done before timeout. Callback exceptionally when the future
   * completed exceptionally or future isn't done after timeout.
   *
   * @param future
   *   The future that needs to be watched.
   * @param timeout
   *   timeout
   * @param unit
   *   timeout unit
   * @param handle
   *   The handler will be called when future complete normally before timeout.
   * @param exceptionally
   *   The exceptionally will be called when future completed exceptionally or future isn't done
   *   after timeout.
   */
  def runTimeout[T](
      future: CompletableFuture[T],
      timeout: Long,
      unit: TimeUnit,
      handle: Consumer[T],
      exceptionally: Consumer[Throwable]): CompletableFuture[Unit] = {

    future
      .applyToEither(
        setTimeout(timeout, unit),
        new JavaFunc[T, Unit]() {
          override def apply(t: T): Unit = {
            if (handle != null) {
              handle.accept(t)
            }
          }
        })
      .exceptionally(new JavaFunc[Throwable, Unit]() {
        override def apply(t: Throwable): Unit = {
          if (exceptionally != null) {
            exceptionally.accept(t)
          }
        }
      })
      .whenComplete(new BiConsumer[Unit, Throwable]() {
        override def accept(t: Unit, u: Throwable): Unit = {
          if (!future.isDone) {
            future.cancel(true)
          }
        }
      })
  }

  def runTimeout[T](
      future: CompletableFuture[T],
      timeout: Long,
      unit: TimeUnit): CompletableFuture[Unit] = {
    runTimeout(future, timeout, unit, null, null)
  }

}

File: streampark-common/src/main/java/org/apache/streampark/common/util/CompletableFutureUtils.java
/*
 * Licensed to the Apache Software Foundation (ASF) under one or more
 * contributor license agreements.  See the NOTICE file distributed with
 * this work for additional information regarding copyright ownership.
 * The ASF licenses this file to You under the Apache License, Version 2.0
 * (the "License"); you may not use this file except in compliance with
 * the License.  You may obtain a copy of the License at
 *
 *    http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 */

package org.apache.streampark.common.util;

import java.util.concurrent.Callable;
import java.util.concurrent.CompletableFuture;
import java.util.concurrent.ScheduledThreadPoolExecutor;
import java.util.concurrent.ThreadFactory;
import java.util.concurrent.TimeUnit;
import java.util.concurrent.TimeoutException;
import java.util.function.BiConsumer;
import java.util.function.Consumer;
import java.util.function.Function;

public final class CompletableFutureUtils {

    private CompletableFutureUtils() {
    }

    private static class DaemonThreadFactory implements ThreadFactory {

        @Override
        public Thread newThread(Runnable r) {
            Thread t = new Thread(r);
            t.setDaemon(true);
            t.setName("CompletableFutureDelayScheduler");
            return t;
        }
    }

    private static final ScheduledThreadPoolExecutor COMPLETABLE_DELAYER;

    static {
        COMPLETABLE_DELAYER = new ScheduledThreadPoolExecutor(1, new DaemonThreadFactory());
        COMPLETABLE_DELAYER.setRemoveOnCancelPolicy(true);
    }

    private static <T> CompletableFuture<T> setTimeout(long timeout, TimeUnit unit) {
        CompletableFuture<T> result = new CompletableFuture<>();
        COMPLETABLE_DELAYER.schedule(
            (Callable<Boolean>) () -> {
                result.completeExceptionally(new TimeoutException());
                return true;
            },
            timeout,
            unit);
        return result;
    }

    public static <T> CompletableFuture<T> supplyTimeout(
                                                         CompletableFuture<T> future,
                                                         long timeout,
                                                         TimeUnit unit,
                                                         Function<T, T> handle,
                                                         Function<Throwable, T> exceptionally) {
        return future.applyToEither(setTimeout(timeout, unit), handle)
            .exceptionally(exceptionally)
            .whenComplete(
                (BiConsumer<T, Throwable>) (t, u) -> {
                    if (!future.isDone()) {
                        future.cancel(true);
                    }
                });
    }

    public static <T> CompletableFuture<Void> runTimeout(
                                                         CompletableFuture<T> future, long timeout, TimeUnit unit) {
        return runTimeout(future, timeout, unit, null, null);
    }

    public static <T> CompletableFuture<Void> runTimeout(
                                                         CompletableFuture<T> future,
                                                         long timeout,
                                                         TimeUnit unit,
                                                         Consumer<T> handle,
                                                         Consumer<Throwable> exceptionally) {
        return future.applyToEither(
            setTimeout(timeout, unit),
            (Function<T, Void>) t -> {
                if (handle != null) {
                    handle.accept(t);
                }
                return null;
            })
            .exceptionally(
                (Function<Throwable, Void>) t -> {
                    if (exceptionally != null) {
                        exceptionally.accept(t);
                    }
                    return null;
                })
            .whenComplete(
                (BiConsumer<Void, Throwable>) (t, u) -> {
                    if (!future.isDone()) {
                        future.cancel(true);
                    }
                });
    }
}
